theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -334,18 +331,17 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
     }
   }
 
-  if (params->choose_algo && (!params->choose_once || !reuse_algo)) {
-    // algo may have changed due to fallback, we must update it.
+  if (params->choose_algo && !reuse_algo) {
+    // save for next time/cache
     prev_algo.algo = algo;
-    // save worksize for next time/cache
     prev_algo.wsSize = worksize;
-
-    // Add to the cache if we choose on shape change, or first time if we choose once.
-    dnn_conv_update_cache(hashkey, prev_algo);
-  }
+    prev_algo.mathType = mathtype;
+    // Add to the cache if we choose on shape change, or first time if
+    // we choose once.
+    if (!use_cached)
+      dnn_conv_update_cache(hashkey, prev_algo);
 
 #ifdef DEBUG
```

Comment:
With these changes, this debug code is not printed anymore when an algorithm is reused (for time_once or guess_once), as this code is now in a block with condition `if params->choose_algo and not reused`. So, it will not help us debug time_once and guess_once algorithms.

Same remark for dnn_gw and dnn_gi.